From 04ecf33b0db0a32fc4676b63aab530f112838926 Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <natechancellor@gmail.com>
Date: Fri, 31 Aug 2018 19:15:38 -0700
Subject: [PATCH 2/2] staging: ion: msm: Apply insert-commit-id-here to
 msm_ion_custom_ioctl

Commit insert-commit-id-here ("staging: android: ion: fix ION_IOC_{MAP,SHARE}
use-after-free") removed the definition of ion_handle_get_by_id,
replacing it with a lockless version. Update msm_ion_custom_ioctl with
this new definition and locking scheme for completeness (it appears that
the ioctl does not suffer from the issue outlined by Greg Hackmann).

Signed-off-by: Nathan Chancellor <natechancellor@gmail.com>
---
 drivers/staging/android/ion/ion.c         | 2 +-
 drivers/staging/android/ion/ion_priv.h    | 4 ++--
 drivers/staging/android/ion/msm/msm_ion.c | 5 ++++-
 3 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/android/ion/ion.c b/drivers/staging/android/ion/ion.c
index 18187d8a58ca..d0479bf28791 100644
--- a/drivers/staging/android/ion/ion.c
+++ b/drivers/staging/android/ion/ion.c
@@ -420,7 +420,7 @@ static struct ion_handle *ion_handle_lookup(struct ion_client *client,
 	return ERR_PTR(-EINVAL);
 }
 
-static struct ion_handle *ion_handle_get_by_id_nolock(struct ion_client *client,
+struct ion_handle *ion_handle_get_by_id_nolock(struct ion_client *client,
 						int id)
 {
 	struct ion_handle *handle;
diff --git a/drivers/staging/android/ion/ion_priv.h b/drivers/staging/android/ion/ion_priv.h
index c466df94c874..90092e4fe242 100644
--- a/drivers/staging/android/ion/ion_priv.h
+++ b/drivers/staging/android/ion/ion_priv.h
@@ -581,8 +581,8 @@ int ion_walk_heaps(struct ion_client *client, int heap_id,
 			unsigned int type, void *data,
 			int (*f)(struct ion_heap *heap, void *data));
 
-struct ion_handle *ion_handle_get_by_id(struct ion_client *client,
-					int id);
+struct ion_handle *ion_handle_get_by_id_nolock(struct ion_client *client,
+						int id);
 
 int ion_handle_put(struct ion_handle *handle);
 
diff --git a/drivers/staging/android/ion/msm/msm_ion.c b/drivers/staging/android/ion/msm/msm_ion.c
index 518f660c24e1..05ebc2217a8c 100644
--- a/drivers/staging/android/ion/msm/msm_ion.c
+++ b/drivers/staging/android/ion/msm/msm_ion.c
@@ -725,13 +725,16 @@ long msm_ion_custom_ioctl(struct ion_client *client,
 		struct mm_struct *mm = current->active_mm;
 
 		if (data.flush_data.handle > 0) {
-			handle = ion_handle_get_by_id(client,
+			mutex_lock(&client->lock);
+			handle = ion_handle_get_by_id_nolock(client,
 						(int)data.flush_data.handle);
 			if (IS_ERR(handle)) {
+				mutex_unlock(&client->lock);
 				pr_info("%s: Could not find handle: %d\n",
 					__func__, (int)data.flush_data.handle);
 				return PTR_ERR(handle);
 			}
+			mutex_unlock(&client->lock);
 		} else {
 			handle = ion_import_dma_buf(client, data.flush_data.fd);
 			if (IS_ERR(handle)) {
-- 
2.18.0

